cmd/snap/cmd_run.go

```diff
@@ -91,6 +93,40 @@ and environment.
 		}, nil)
 }
 
+func systemKeyMismatch() (bool, error) {
+	raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
+	if err != nil && !os.IsNotExist(err) {
+		return false, err
+	}
+	systemKey := string(raw)
+	return systemKey != interfaces.SystemKey(), nil
```

Comment:
This means that snap run must be able to compute the system key which is non-trivial in complexity. 

How about this pattern:

- we look at /proc/uptime
- we stat the system key file
- if the system key is more recent than time since boot time then we are good and there's nothing to do
- if the system key is older then assume it is out of date and wait

On snapd side:
- if the system key is correct then open it and update mtime
- in all other cases behave as before.

This way snap run can easily and quickly check that we are in a good state without involving RPC or really high complexity operations.

Reply:
I think we also have a race with classic when it restarts so I think it is not sufficient to look at uptime. Also (just for the record): the system-key is not trivial but its also not expensive, it does not involve RPC or highly complex operations, it does: look at the build-id (open /proc/self/exe and reading the elf header), reading the apparmor sysfs dir, checking for nfs (reading the mountinfo file), checking if root is on an overlay (reading the mountinfo file), checking core revision (one readlink call), reading the seccomp actions (one read in proc which will happen on startup anyway). So I agree there is overhead but I would not call it highly complex :)